hazelcast/src/test/java/com/hazelcast/jet/core/JetTestSupport.java

```diff
@@ -87,10 +100,61 @@
     @After
     public void shutdownFactory() throws Exception {
         if (instanceFactory != null) {
+            Map<Long, String> leakedClassloaders = shutdownJobsAndGetLeakedClassLoaders();
+
             SUPPORT_LOGGER.info("Terminating instanceFactory in JetTestSupport.@After");
             spawn(() -> instanceFactory.terminateAll())
                     .get(1, TimeUnit.MINUTES);
+
+            if (!leakedClassloaders.isEmpty()) {
+                String ids = leakedClassloaders
+                        .entrySet().stream()
+                        .map(entry -> idToString(entry.getKey()) + "[" + entry.getValue() + "]")
+                        .collect(joining(", "));
+                fail("There are one or more leaked job classloaders. " +
+                        "This is a bug, but it is not necessarily related to this test. " +
+                        "The classloader was leaked for the following jobIds: " + ids);
+            }
+        }
+    }
+
+    @Nonnull
+    private Map<Long, String> shutdownJobsAndGetLeakedClassLoaders() {
+        Map<Long, String> leakedClassloaders = new HashMap<>();
+        Collection<HazelcastInstance> instances = instanceFactory.getAllHazelcastInstances();
+        for (HazelcastInstance instance : instances) {
+            if (instance.getConfig().getJetConfig().isEnabled()) {
+                // Some tests leave jobs running, which keeps job classloader, shut down all running/starting jobs
+                JetService jet = instance.getJet();
+                List<Job> jobs = jet.getJobs();
+                for (Job job : jobs) {
+                    if (job.getStatus() == RUNNING || job.getStatus() == STARTING) {
+                        job.cancel();
+                        assertJobStatusEventually(job, FAILED); // Failed = cancelled
+                    }
+                }
+
+                JobClassLoaderService jobClassLoaderService = ((HazelcastInstanceImpl) instance).node
+                        .getNodeEngine()
+                        .<JetServiceBackend>getService(SERVICE_NAME)
+                        .getJobClassLoaderService();
+
+                ConcurrentHashMap<Long, ?> classLoaders = readFieldOrNull(jobClassLoaderService, "classLoaders");
```

Comment:
Why don't we expose the field publicly? You're accessing it anyway and at least it's visible in code navigation.